### PR TITLE
Switching Lease to async methods where needed

### PR DESCRIPTION
**Pull request by SeanFeldman**
Started with 6 failing tests, not sure if those are _just_ flaky or not.
Please review.

### CODE REVIEW

File: src/Microsoft.Azure.EventHubs.Processor/AzureBlobLease.cs [hunk 1]
```diff
@@ -3,83 +3,68 @@
...
-        }
+		public override async Task<bool> IsExpired()
+		{
+			await this.Blob.FetchAttributesAsync().ConfigureAwait(false); // Get the latest metadata
+			var currentState = this.Blob.Properties.LeaseState;
+			return currentState != LeaseState.Leased;
+		}
 
-        public long SequenceNumber { get; set; }
+		public override async Task<string> GetStateDebug()
```
Comment (serkantkaraca) [inline src/Microsoft.Azure.EventHubs.Processor/AzureBlobLease.cs]: >public override async Task<string> GetStateDebug() [](start = 2, length = 50)

Can you please remove this as part of you change? It seems some debug code leaked.

**Comment by serkantkaraca** [comment]
Sean, what tests are failing? We don't have any flaky tests AFAIK. I have been running them for a while and never saw more than 1 test failing at a full run.

File: src/Microsoft.Azure.EventHubs.Processor/Lease.cs [hunk 2]
```diff
@@ -35,15 +34,15 @@ protected Lease(Lease source)
...
-        public virtual bool IsExpired() 
+        public virtual Task<bool> IsExpired() 
         {
     	    // this function is meaningless in the base class
-    	    return false;
+            return Task.FromResult(false);
         }
 
-        public virtual string GetStateDebug()
+        public virtual Task<string> GetStateDebug()
```
Comment (SeanFeldman) [inline src/Microsoft.Azure.EventHubs.Processor/Lease.cs]: @serkantkaraca would it add value to decorate with [`DebuggerDisplayAttribute`](https://msdn.microsoft.com/en-us/library/ms228992(v=vs.110).aspx)? I see the value in this, so perhaps `DebuggerTypeProxy` could do the job and leave public API cleaner?

**Comment by SeanFeldman** [comment]
@serkantkaraca both @jtaubensee and I had failing tests. I'll list them a little later.

**Comment by serkantkaraca** [comment]
Sean, you might be using an invalid lease container name. Can you sync and rerun the tests? I have just merged a change to validate lease container name provided and throw accordingly.

**Comment by serkantkaraca** [comment]
Sean, you might be using an invalid lease container name. Can you sync and rerun the tests? I have just merged a change to validate lease container name provided and throw accordingly.

**Comment by SeanFeldman** [comment]
@serkantkaraca for the failing tests, I've raised a separate issue (#14) as it looks unrelated to this PR.

**Comment by SeanFeldman** [comment]
@serkantkaraca removed `GetStateDebug()` as requested.

**Comment by serkantkaraca** [comment]
:shipit:

**Comment by SeanFeldman** [comment]
Awesome! Thank you @serkantkaraca 🎆

**Pull request merged**